src/analyses/call_graph.cpp

```diff
@@ -51,50 +46,158 @@ void call_grapht::add(
 {
-  out << "digraph call_graph {\n";
-
-  for(const auto &edge : graph)
+  std::list<irep_idt> result;
+  node_indext src_idx, dest_idx;
+  if(!get_node_index(src, src_idx))
+    throw "unable to find src function in call graph";
+  if(!get_node_index(dest, dest_idx))
+    throw "unable to find destination function in call graph";
+
+  patht path;
+  shortest_path(src_idx, dest_idx, path);
+  for(const auto &n : path)
   {
-    out << "  \"" << edge.first << "\" -> "
-        << "\"" << edge.second << "\" "
-        << " [arrowhead=\"vee\"];"
-        << "\n";
+    result.push_back(nodes[n].function_name);
   }
-
-  out << "}\n";
+  return result;
 }
 
-void call_grapht::output(std::ostream &out) const
+void call_grapht::reachable_within_n_steps(
+    std::size_t steps,
+    std::unordered_set<irep_idt,
+    irep_id_hash> & function_list)
 {
-  for(const auto &edge : graph)
+  std::list<node_indext> worklist;
+
+  for(const auto &f : function_list)
   {
-    out << edge.first << " -> " << edge.second << "\n";
+    node_indext start_index;
+    if(get_node_index(f, start_index))
+      worklist.push_back(start_index);
+    else
+      throw "function not found in call graph";
+  }
+
+  // mark end of level 0
+  worklist.push_back(std::numeric_limits<std::size_t>::max());
+  std::size_t depth=0;
+
+  while(!worklist.empty())
+  {
+    const node_indext id = worklist.front();
+    worklist.pop_front();
+
+    // check if we have hit end of level
+    if(id == std::numeric_limits<std::size_t>::max())
+    {
+      depth++;
+      // mark end of next level
+      if(!worklist.empty())
+        worklist.push_back(id);
+      continue;
+    }
+    function_list.insert(nodes[id].function_name);
+
+    if(depth < steps)
+    {
+      for(const auto &o : nodes[id].out)
+      {
+        if(function_list.find(nodes[o.first].function_name)
+            == function_list.end())
+          worklist.push_back(o.first);
+      }
+    }
   }
 }
 
-void call_grapht::output_xml(std::ostream &out) const
+std::unordered_set<irep_idt, irep_id_hash>
+call_grapht::reachable_functions(irep_idt start_function)
 {
-  for(const auto &edge : graph)
+  std::unordered_set<irep_idt, irep_id_hash> result;
+  std::list<node_indext> worklist;
+  node_indext start_index;
+
+  if(get_node_index(start_function, start_index))
+    worklist.push_back(start_index);
+  else
+    throw "no start function found in call graph";
+
+  while(!worklist.empty())
   {
-    out << "<call_graph_edge caller=\"";
-    xmlt::escape_attribute(id2string(edge.first), out);
-    out << "\" callee=\"";
-    xmlt::escape_attribute(id2string(edge.second), out);
-    out << "\">\n";
+    const node_indext id = worklist.front();
+    worklist.pop_front();
+
+    result.insert(nodes[id].function_name);
+    for(const auto &o : nodes[id].out)
+    {
+      if(result.find(nodes[o.first].function_name) == result.end())
+        worklist.push_back(o.first);
+    }
   }
+
+  return result;
 }
```

Comment:
How much work would it be to move these algorithms out into util/graph.h.  It seems like this might not be the only place we want to do a shortest path or bounded depth search.

Reply:
There is already a branch by Smowton that combines my call graph pull request and his, so I think I should probably incorporate this change into that.

Reply:
OK.  If the two of you get together on this I'd be happy to review.